### ISSUE TITLE
"Go To Definition" on re-abstraction doesn't go the re-abstracted member

### ISSUE THREAD

**Issue by AlekseyTs** [dotnet/roslyn#61268]
```
interface I1<T1> where T1 : I1<T1>
{
    static abstract void M1();
    void M3();
}

interface I3<T3> : I1<T3> where T3 : I3<T3>
{
    static abstract void I1<T3>.M1();
    abstract void I1<T3>.M3();
}
```

Try using "Go To Definition" on abstract explicit implementations.

Observed:
Doesn't work.

Expected:
Should go to the corresponding definition in I1.

**Comment by AlekseyTs** [comment]
CC @sharwell

**Comment by CyrusNajmabadi** [comment]
@AlekseyTs  I debugged into this on the IDE side.  We make a call here to determine what interface members a particular member is an implementation of:

```c#
    public static ImmutableArray<ISymbol> ExplicitOrImplicitInterfaceImplementations(this ISymbol symbol)
    {
        if (symbol.Kind is not SymbolKind.Method and not SymbolKind.Property and not SymbolKind.Event)
            return [];

        var containingType = symbol.ContainingType;
        var query = from iface in containingType.AllInterfaces
                    from interfaceMember in iface.GetMembers()
                    let impl = containingType.FindImplementationForInterfaceMember(interfaceMember)
                    where symbol.Equals(impl)
                    select interfaceMember;
        return query.ToImmutableArray();
    }
```

However, in the bowels of FindImplementationForInterfaceMember it eventually gets to: TypeSymbol.FindMostSpecificImplementation which does this:

```c#
                case 1:
                    {
                        Symbol result = implementingMember.Single();

                        if (result.IsAbstract)
                        {
                            return null;
                        }
```

As this member is abstract, it's  not considered an implementation of the interface member. 

Is this the expected behavior of this API?   If so, do you have a recommendation on how to get the originating interface member that an explicit abstract member is an implementation of?

**Comment by AlekseyTs** [comment]
@CyrusNajmabadi I am not sure why ```FindImplementationForInterfaceMember``` is used in the specific scenario. I expect ```ExplicitInterfaceImplementations``` API on an explicit interface implementation to point directly to the method in the interface.  

> Is this the expected behavior of this API?

I think this behavior is expected. The re-abstraction is effectively erases an implementation if one was inherited from base interfaces. Which means that the interface member is not implemented, i.e. doesn't have an implementation. The purpose of ```FindImplementationForInterfaceMember``` is to find implementation, if an interface member has one.

**Comment by AlekseyTs** [comment]
> Do you have any suggestions on how we can figure out this mapping? Is there something else exposed in the symbol api to link these up?

I already mentioned ```ExplicitInterfaceImplementations```. If one wants to know what an "an explicit interface implementation" member implements,  ```ExplicitInterfaceImplementations``` API answers this question.

**Comment by CyrusNajmabadi** [comment]
Oh sorry, i inserted a 'not' in my head and misread your post.  Let me try that out :)

**Comment by CyrusNajmabadi** [comment]
So this is still bit weird.  As you mention:  Which means that the interface member is not implemented, i.e. doesn't have an implementation.

And yet "ExplicitInterfaceImplementations" *does* include it... which seems to imply that it *is* an implementation.   The docs for "ExplicitInterfaceImplementation" even say:

```
        /// <summary>
        /// Returns interface methods explicitly implemented by this method.
        /// </summary>
```

Which indicates it's *explicitly implemented*.   Finally the docs for `FindImplementationForInterfaceMember` says:

> Returns the corresponding symbol in this type or a base type that implements interfaceMember (either implicitly *or explicitly*)

The docs for this say that it will return a symbol that "implements interface member (... explicitly)".

And this member doesn't indicate that it explicitly implements the interface member (through ExplicitInterfaceImplementation).  So it's unclear to me why this doesn't work.  

Can you clarify what about this case makes the member no longer an "implementation" from the perspective of  "FindImplementationForInterfaceMember" but still is an "implementation" from the perspective of "ExplicitInterfaceImplementation"?  I've read your explanation a few times, but it's not clear why it makes the current  symbol model fall out:

> The re-abstraction is effectively erases an implementation if one was inherited from base interfaces. Which means that the interface member is not implemented, i.e. doesn't have an implementation

If the "the interface member is not implemented, i.e. doesn't have an implementation", then i would not expect ExplicitInterfaceImplementations to return anything, since the it "doesn't have an implementation". 

Thanks!

**Comment by AlekseyTs** [comment]
For better or worse, an explicit interface implementation syntax is used for re-abstraction. Re-abstraction doesn't implement an interface members, it erases one. I understand that one might find this confusing. ```FindImplementationForInterfaceMember``` is meant to find an implementation that actually implements the member from runtime and language perspective. I.e. something that an entry in a v-table points to, a method to call when an interface method is invoked. Re-abstraction is not that. Its the opposite, it clears that slot in a v-table. If a type implements a member, derived type doesn't have to implement it (it is considered implemented). If a member is re-abstracted, it is not implemented, a derived type must implement it, hence the behavior of ```FindImplementationForInterfaceMember```. 

In other words, a re-abstraction is a syntactic form of explicit implementation that "implements" a member (hence ```ExplicitInterfaceImplementation``` behavior) in such a way that the member is treated as not implemented.

I am happy to continue this discussion offline.

**Comment by CyrusNajmabadi** [comment]
Great.  That helps.  I'd recommend doc'ing this then though as the current docs def seem misleading.  Thanks :)
